src/world/Units/Creatures/AIInterface.cpp

```diff
@@ -2477,3 +3091,546 @@ uint32 AIInterface::fixupCorridor(dtPolyRef* path, const uint32 npath, const uin
+{
+    mMaxCount = castCount;
+}
+
+uint32_t CreatureAISpells::getMaxCastCount()
+{
+    return mMaxCount;
+}
+
+uint32_t CreatureAISpells::getCastCount()
+{
+    return mCastCount;
+}
+
+bool CreatureAISpells::isDistanceInRange(float targetDistance)
+{
+    if (targetDistance >= mMinPositionRangeToCast && targetDistance <= mMaxPositionRangeToCast)
+        return true;
+
+    return false;
+}
+
+void CreatureAISpells::setMinMaxDistance(float minDistance, float maxDistance)
+{
+    mMinPositionRangeToCast = minDistance;
+    mMaxPositionRangeToCast = maxDistance;
+}
+
+bool CreatureAISpells::isHpInPercentRange(int targetHp)
+{
+    if (targetHp >= mMinHpRangeToCast && targetHp <= mMaxHpRangeToCast)
+        return true;
+
+    return false;
+}
+
+void CreatureAISpells::setMinMaxPercentHp(int minHp, int maxHp)
+{
+    mMinHpRangeToCast = minHp;
+    mMaxHpRangeToCast = maxHp;
+}
+
+void CreatureAISpells::setAvailableForScriptPhase(std::vector<uint32_t> phaseVector)
+{
+    for (const auto& phase : phaseVector)
+    {
+        mPhaseList.push_back(phase);
+    }
+}
+
+bool CreatureAISpells::isAvailableForScriptPhase(uint32_t scriptPhase)
+{
+    if (mPhaseList.empty())
+        return true;
+
+    for (const auto& availablePhase : mPhaseList)
+    {
+        if (availablePhase == scriptPhase)
+            return true;
+    }
+
+    return false;
+}
+
+void CreatureAISpells::setAttackStopTimer(uint32_t attackStopTime)
+{
+    mAttackStopTimer = attackStopTime;
+}
+
+uint32_t CreatureAISpells::getAttackStopTimer()
+{
+    return mAttackStopTimer;
+}
+
+void CreatureAISpells::setAnnouncement(std::string announcement)
+{
+    mAnnouncement = announcement;
+}
+
+void CreatureAISpells::sendAnnouncement(Unit* pUnit)
+{
+    if (!mAnnouncement.empty() && pUnit != nullptr)
+    {
+        sLogger.debug("AISpellEmotes::sendAnnouncement() : called");
+
+        pUnit->sendChatMessage(CHAT_MSG_RAID_BOSS_EMOTE, LANG_UNIVERSAL, mAnnouncement.c_str());
+    }
+}
+
+void CreatureAISpells::setCustomTarget(Unit* targetCreature)
+{
+    mCustomTargetCreature = targetCreature;
+}
+
+Unit* CreatureAISpells::getCustomTarget()
+{
+    return mCustomTargetCreature;
+}
+
+CreatureAISpells* AIInterface::addAISpell(uint32_t spellId, float castChance, uint32_t targetType, uint32_t duration /*= 0*/, uint32_t cooldown /*= 0*/, bool forceRemove /*= false*/, bool isTriggered /*= false*/)
+{
+    const auto spellInfo = sSpellMgr.getSpellInfo(spellId);
+    if (spellInfo != nullptr)
+    {
+        uint32_t spellDuration = duration * 1000;
+        if (spellDuration == 0)
+            spellDuration = spellInfo->getSpellDefaultDuration(nullptr);
+
+        uint32_t spellCooldown = cooldown * 1000;
+        if (spellCooldown == 0)
+            spellCooldown = spellInfo->getSpellDefaultDuration(nullptr);
+
+        CreatureAISpells* newAISpell = new CreatureAISpells(spellInfo, castChance, targetType, spellDuration, spellCooldown, forceRemove, isTriggered);
```

Comment:
don't forget to delete it ;-) it is not enough to clear the storage  on shutdown you have to delete every spointer you created with new at least in the destructor.

Reply:
i return the pointer on that function so how would i delete it and still return it?